### PR TITLE
Feature flowchart documentation

### PR DESCRIPTION
**Pull request by bcouston**
Discussion on certificate questionnaire flowchart functionality.

### CODE REVIEW

**Comment by pkqk** [comment]
@quadrophobiac You removed the `config/database.example.yml` file in a90433bb05c46f320b1f01289096a10e750047bc this is used by travis to set up the database config in testing.

File: features/step_definitions/flowchart_steps.rb [hunk 1]
```diff
@@ -0,0 +1,404 @@
+# encoding: utf-8
+require 'pry'
+
+legal_markdown = <<EOF
```
Comment (pezholio) [inline features/step_definitions/flowchart_steps.rb]: I'd probably put this stuff in a text file in /fixtures, and load it in the steps, just to keep this file cleaner

File: lib/flow.rb [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
     @xml_copy = Nokogiri::XML(File.open(path))
+    # marshals XML into two above hashes
+
   end
 
   def get_path(jurisdiction, type)
+    # return two different paths based on whether legal or practical question flow selected
     jurisdiction = "en" if (jurisdiction == "gb" && type == "Practical")
-    jurisdiction.upcase! if type == "Legal"
+    jurisdiction.upcase! if type == "Legal" # check what this bang! means
```
Comment (pezholio) [inline lib/flow.rb]: This performs the transformation in place, so we don't have to do jurisdiction = jurisdiction.upcase

File: app/controllers/flowcharts_controller.rb [hunk 3]
```diff
@@ -1,19 +1,45 @@
...
     flow = Flow.new(@jurisdiction, @type)
-    @questions = flow.questions
+    @questions = flow.questions # returns an array of hashes with dependency always nil
     @dependencies = flow.dependencies
     @deps = []
+    # array of string containing a rendered partial
     @levels = {}
-    (@questions + @dependencies).each do |q|
+
+    (@questions + @dependencies).each do |q| #what operation is this '+'
```
Comment (pezholio) [inline app/controllers/flowcharts_controller.rb]: This puts both `@questions` and `@dependencies` in one array

**Comment by bcouston** [comment]
Added issue #1216 for missing questions in the Legal flowchart.

**Comment by bcouston** [comment]
So far @quadrophobiac and I have written tests for creating a flow chart, counting the number of questions, dependencies and answers, one to do with prerequisites and one that tests all the halting conditions of the flowchart when an end block is displayed. At the moment for questions, dependencies and answers, the expected amount is a custom amount set by manually looking at the xml. After this proposed change #1217 , I can draw an expected value from the number of elements in the model and compare it to what's outputted by the helper. Also this change will hopefully provide a suitable fix for #1216. @pezholio , do you have any more ideas for things to test?

**Comment by bcouston** [comment]
There's also a lot of errors of this nature approximately 25, I'm not sure if they've been triggered by something we've changed while working on flowcharts.
image

**Comment by pkqk** [comment]
@bcouston well all other branches are passing so it has to be something on this branch that is different.

You can use `git bisect` to narrow down which commits broke certain things

**Comment by pezholio** [comment]
I think that's the lot test wise really. The only remaining thing I guess is what we discussed yesterday with the `!== ''` dependencies. There's two ways we can tackle this:

* We check if an input has a dependency of this nature, if it does, we add that question directly after (if a user has chosen to leave that input blank, then they can assume that they leave that one blank too)
* If a question has they type `input`, we treat it as a multiple choice question with two answers, if the answer is blank, the dependency is `nil`, but if the answer has content, then we go to the dependency (which we'll need to calculate by looking for `//if[contains(@test, this.questionID() !== ')]'` (or similar))

Does this make sense? Sorry, this has taken half a day to write :smile:

**Comment by bcouston** [comment]
Great, yes @pezholio I do understand. I think another solution would be implementing #1217, where a text input answer is recognised and see if that fixes #1216?

**Comment by Floppy** [comment]
I'm having a quick look at the test failures and where it's come from...

**Pull request merged**